### PR TITLE
[CORDA-1799]: Avoid generating test CorDapp JARs from each out of process node started by the driver

### PR DESCRIPTION
**Pull request by sollecitom**
https://r3-cev.atlassian.net/browse/CORDA-1799

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/cordapp/Cordapp.kt [hunk 1]
```diff
@@ -48,4 +48,20 @@ interface Cordapp {
...
+
+    /**
+     * CorDapp's information, including vendor and version.
+     *
+     * @property shortName Cordapp's shortName
+     * @property vendor Cordapp's vendor
+     * @property version Cordapp's version
+     */
+    @DoNotImplement
+    interface Info {
```
Comment (sollecitom) [inline core/src/main/kotlin/net/corda/core/cordapp/Cordapp.kt]: Carried across from ent.

File: core/src/main/kotlin/net/corda/core/internal/cordapp/CordappImpl.kt [hunk 2]
```diff
@@ -33,4 +33,16 @@ data class CordappImpl(
      * TODO: Also add [SchedulableFlow] as a Cordapp class
      */
     override val cordappClasses = ((rpcFlows + initiatedFlows + services + serializationWhitelists.map { javaClass }).map { it.name } + contractClassNames)
+
+    data class Info(override val shortName: String, override val vendor: String, override val version: String): Cordapp.Info {
```
Comment (sollecitom) [inline core/src/main/kotlin/net/corda/core/internal/cordapp/CordappImpl.kt]: Carried across from ent.

File: .ci/api-current.txt [hunk 3]
```diff
@@ -6297,14 +6297,14 @@ public final class net.corda.testing.node.MockNodeParameters extends java.lang.O
...
   public <init>()
-  public <init>(java.util.List<String>)
-  public <init>(java.util.List<String>, net.corda.core.identity.CordaX500Name)
-  public <init>(java.util.List<String>, net.corda.core.identity.CordaX500Name, java.security.KeyPair, java.security.KeyPair...)
-  public <init>(java.util.List<String>, net.corda.core.identity.CordaX500Name, net.corda.core.node.services.IdentityService)
-  public <init>(java.util.List<String>, net.corda.core.identity.CordaX500Name, net.corda.core.node.services.IdentityService, java.security.KeyPair, java.security.KeyPair...)
-  public <init>(java.util.List<String>, net.corda.testing.core.TestIdentity, net.corda.core.node.services.IdentityService, net.corda.core.node.NetworkParameters, java.security.KeyPair...)
-  public <init>(java.util.List<String>, net.corda.testing.core.TestIdentity, net.corda.core.node.services.IdentityService, java.security.KeyPair...)
-  public <init>(java.util.List<String>, net.corda.testing.core.TestIdentity, java.security.KeyPair...)
+  public <init>(Iterable<String>)
```
Comment (tlil) [inline .ci/api-current.txt]: You reverted this, no? Update the api-current.txt again?

File: docs/source/changelog.rst [hunk 4]
```diff
@@ -6,6 +6,8 @@ release, see :doc:`upgrade-notes`.
 
 Unreleased
 ----------
+* Introduced ``TestCorDapp`` and utilities to support asymmetric setups for nodes through ``DriverDSL``, ``MockNetwork`` and ``MockServices``.
```
Comment (tlil) [inline docs/source/changelog.rst]: Anything that needs to go into the `upgrade-notes.rst`?

File: node/src/integration-test/kotlin/net/corda/node/flows/AsymmetricCorDappsTests.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            }
+        }
+    }
+
+    @Test
+    fun noSharedCorDappsWithAsymmetricSpecificClasses() {
+
+        driver(DriverParameters(startNodesInProcess = false, cordappsForAllNodes = emptySet())) {
+
+            val nodeA = startNode(additionalCordapps = setOf(TestCorDapp.Factory.create("Szymon CorDapp", "1.0", classes = setOf(Ping::class.java)))).getOrThrow()
```
Comment (tlil) [inline node/src/integration-test/kotlin/net/corda/node/flows/AsymmetricCorDappsTests.kt]: This is neat 👍

Why an explicit `Factory` inner class on `TestCorDapp` btw?

File: .ci/api-current.txt (hunk 3 above)
Comment (sollecitom) [reply .ci/api-current.txt]: No, it's still there, but I replaced `java.util.List` with the incorrect `java.util.Iterable` initially.

File: build.gradle [hunk 6]
```diff
@@ -168,7 +168,7 @@ allprojects {
             jvmTarget = "1.8"
             javaParameters = true   // Useful for reflection.
             freeCompilerArgs = ['-Xjvm-default=compatibility']
-            allWarningsAsErrors = true
+            allWarningsAsErrors = false
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply build.gradle]: Yes, I'll remove this as soon as I'm done locally (we should remove it for sure, but as a separate PR).

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/InternalMockNetwork.kt [hunk 7]
```diff
@@ -86,25 +81,26 @@ data class InternalMockNodeParameters(
...
             mockNodeParameters.configOverrides,
             MOCK_VERSION_INFO,
-            mockNodeParameters.extraCordappPackages
+            mockNodeParameters.additionalCordapps
     )
 }
 
-open class InternalMockNetwork(private val cordappPackages: List<String> = emptyList(),
-                               defaultParameters: MockNetworkParameters = MockNetworkParameters(),
+open class InternalMockNetwork(defaultParameters: MockNetworkParameters = MockNetworkParameters(),
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/InternalMockNetwork.kt]: InternalMockNetwork is for us, and I've already changed it everywhere.

File: core/src/main/kotlin/net/corda/core/internal/cordapp/CordappImpl.kt [hunk 8]
```diff
@@ -33,4 +33,16 @@ data class CordappImpl(
...
+        companion object {
+            private const val UNKNOWN_VALUE = "Unknown"
+
+            val UNKNOWN = Info(UNKNOWN_VALUE, UNKNOWN_VALUE, UNKNOWN_VALUE)
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply core/src/main/kotlin/net/corda/core/internal/cordapp/CordappImpl.kt]: Java perspective? This is internal.

[the remark this answers is not in the document]
Comment (sollecitom) [reply core/src/main/kotlin/net/corda/core/internal/cordapp/CordappImpl.kt]: No no, the interface is fine, this is used by use when reading Cordformation setups.

File: node/src/integration-test/kotlin/net/corda/node/flows/AsymmetricCorDappsTests.kt (hunk 5 above)
Comment (sollecitom) [reply node/src/integration-test/kotlin/net/corda/node/flows/AsymmetricCorDappsTests.kt]: Well, a static factory method doesn't play well from Java apparently.

File: core/src/main/kotlin/net/corda/core/internal/cordapp/CordappImpl.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (tlil) [reply core/src/main/kotlin/net/corda/core/internal/cordapp/CordappImpl.kt]: Ah, alright

File: node/src/integration-test/kotlin/net/corda/node/flows/AsymmetricCorDappsTests.kt (hunk 5 above)
Comment (tlil) [reply node/src/integration-test/kotlin/net/corda/node/flows/AsymmetricCorDappsTests.kt]: Ah, it's an interface. Gotcha

File: node/src/integration-test/kotlin/net/test/cordapp/v1/FlowCheckpointCordapp.kt [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        fun tracker() = ProgressTracker(GENERATING_TRANSACTION, VERIFYING_TRANSACTION, SIGNING_TRANSACTION, FINALISING_TRANSACTION)
+    }
+
+    override val progressTracker = tracker()
+
+    @Suspendable
+    override fun call(): SignedTransaction {
+        progressTracker.currentStep = GENERATING_TRANSACTION
+
+        val messageState = MessageState(message = message, by = /*reciepent ?:*/ ourIdentity)
```
Comment (tlil) [inline node/src/integration-test/kotlin/net/test/cordapp/v1/FlowCheckpointCordapp.kt]: What's this comment doing here? :-)

File: node/src/main/kotlin/net/corda/node/cordapp/CordappLoader.kt [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Handles CorDapp loading
+ */
+interface CordappLoader {
+
+    val cordapps: List<Cordapp>
+
+    val appClassLoader: ClassLoader
+
+    val flowCordappMap: Map<Class<out FlowLogic<*>>, Cordapp>
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/cordapp/CordappLoader.kt]: Some Kdocs for these?

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 11]
```diff
@@ -136,6 +133,9 @@ abstract class PortAllocation {
  * @property startInSameProcess Determines if the node should be started inside the same process the Driver is running
  *     in. If null the Driver-level value will be used.
  * @property maximumHeapSize The maximum JVM heap size to use for the node.
+ * @property logLevel Logging level threshold.
+ * @property additionalCordapps additional [TestCorDapp]s that this node will have available, in addition to the ones common to all nodes managed by the [DriverDSL].
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Capitalise `additional` on this line and `whether` on the next line.

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 12]
```diff
@@ -319,7 +464,41 @@ data class DriverParameters(
...
+            debugPortAllocation,
+            systemProperties,
+            useTestClock,
+            startNodesInProcess,
+            waitForAllNodesToFinish,
+            notarySpecs,
+            extraCordappPackagesToScan,
+            jmxPolicy,
+            networkParameters,
+            emptyMap(),
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]:
```kotlin
notaryCustomOverrides = emptyMap(),
initialiseSerialization = true,
inMemoryDB = true,
```

to make these three lines easier to follow?

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 13]
```diff
@@ -319,7 +464,41 @@ data class DriverParameters(
             networkParameters,
             emptyMap(),
             true,
-            true
+            true,
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]:
```kotlin
notaryCustomOverrides = emptyMap(),
initialiseSerialization = true,
inMemoryDB = true,
```

to make these three lines easier to follow?

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/DriverDSL.kt [hunk 14]
```diff
@@ -80,6 +80,38 @@ interface DriverDSL {
...
+     *     random. Note that this must be unique as the driver uses it as a primary key!
+     * @param rpcUsers List of users who are authorised to use the RPC system. Defaults to empty list.
+     * @param verifierType The type of transaction verifier to use. See: [VerifierType].
+     * @param customOverrides A map of custom node configuration overrides.
+     * @param startInSameProcess Determines if the node should be started inside the same process the Driver is running
+     *     in. If null the Driver-level value will be used.
+     * @param maximumHeapSize The maximum JVM heap size to use for the node as a [String]. By default a number is interpreted
+     *     as being in bytes. Append the letter 'k' or 'K' to the value to indicate Kilobytes, 'm' or 'M' to indicate
+     *     megabytes, and 'g' or 'G' to indicate gigabytes. The default value is "512m" = 512 megabytes.
+     * @param additionalCordapps additional [TestCorDapp]s that this node will have available, in addition to the ones common to all nodes managed by the [DriverDSL].
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/DriverDSL.kt]: Capitalise 'additional' and 'whether'

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/TestCorDapp.kt [hunk 15]
```diff
@@ -0,0 +1,74 @@
+package net.corda.testing.driver
+
+import net.corda.testing.node.internal.MutableTestCorDapp
+import java.net.URL
+import java.nio.file.Path
+
+// TODO add support for CorDapps' config files
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/TestCorDapp.kt]: Will this `TODO` be addressed in this PR?

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/TestCorDapp.kt [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+interface TestCorDapp {
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/TestCorDapp.kt]: Can you add Kdoc's for this one please?

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt [hunk 17]
```diff
@@ -262,7 +287,26 @@ open class MockNetwork(
...
+
+    /**
+     * Create a started node with the given parameters.
+     *
+     * @param legalName The node's legal name.
+     * @param forcedID A unique identifier for the node.
+     * @param entropyRoot The initial entropy value to use when generating keys. Defaults to an (insecure) random value,
+     * but can be overridden to cause nodes to have stable or colliding identity/service keys.
+     * @param configOverrides Add/override behaviour of the [NodeConfiguration] mock object.
+     * @param additionalCordapps additional [TestCorDapp]s that this node will have available, in addition to the ones common to all nodes managed by the [MockNetwork].
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt]: Additional #nit (applies to comments further down as well)

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/MutableTestCorDapp.kt [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+internal class MutableTestCorDapp private constructor(override val name: String, override val version: String, override val vendor: String, override val title: String, private val willResourceBeAddedToCorDapp: (String, URL) -> Boolean, private val jarEntries: Set<JarEntryInfo>) : TestCorDapp.Mutable {
+
+    constructor(name: String, version: String, vendor: String, title: String, classes: Set<Class<*>>, willResourceBeAddedToCorDapp: (String, URL) -> Boolean) : this(name, version, vendor, title, willResourceBeAddedToCorDapp, jarEntriesFromClasses(classes))
+
+    companion object {
+
+        private const val jarExtension = ".jar"
+        private const val whitespace = " "
+        private const val whitespaceReplacement = "_"
+        private val productionPathSegments = setOf<(String) -> String>({ "out${File.separator}production${File.separator}classes" }, { fullyQualifiedName -> "main/${fullyQualifiedName.packageToPath()}" })
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/MutableTestCorDapp.kt]: `Paths.get` + `/` ?

Also the return value of the second one doesn't reference `File.separator` but `/` directly.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/TestCordappsUtils.kt [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+    val manifest = createTestManifest(name, title, version, vendor)
+    return JarOutputStream(outputStream, manifest).use { jos -> zip(jos, willResourceBeAddedBeToCorDapp) }
+}
+
+internal fun Class<*>.jarEntryInfo(): JarEntryInfo {
+
+    return JarEntryInfo.ClassJarEntryInfo(this)
+}
+
+fun Iterable<TestCorDapp>.packageInDirectory(directory: Path) {
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/TestCordappsUtils.kt]: Some doc comments for these ones as well?

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 20]
```diff
@@ -303,6 +309,7 @@ class DriverDSLImpl(
         NON_VALIDATING_BFT(false, CordaX500Name("BFT", "Zurich", "CH"))
     }
 
+    // TODO remove this
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: Will this be done as part of this PR? Or later? (same with the `TODO` on line 370)

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 21]
```diff
@@ -639,10 +658,27 @@ class DriverDSLImpl(
             visibilityHandle.close()
         }
 
-        val useHTTPS = config.typesafe.run { hasPath("useHTTPS") && getBoolean("useHTTPS") }
+        val useHTTPS = specifiedConfig.typesafe.run { hasPath("useHTTPS") && getBoolean("useHTTPS") }
+
+        val existingCorDappDirectoriesOption = if (specifiedConfig.typesafe.hasPath(NodeConfiguration.cordappDirectoriesKey)) specifiedConfig.typesafe.getStringList(NodeConfiguration.cordappDirectoriesKey) else emptyList()
+
+        // TODO caching here doesn't help too much - investigate caching on disk at test or better at test suite level
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: Same re `TODO`s - are you leave this for later?

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 22]
```diff
@@ -145,8 +145,57 @@ data class NodeParameters(
         val customOverrides: Map<String, Any?> = emptyMap(),
         val startInSameProcess: Boolean? = null,
         val maximumHeapSize: String = "512m",
-        val logLevel: String? = null
+        val logLevel: String? = null,
+        val additionalCordapps: Set<TestCorDapp> = emptySet(),
+        val deleteExistingCordappsDirectory: Boolean = false
 ) {
+    /**
+     * Helper builder for configuring a [Node] from Java.
```
Comment (tlil) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: If these helper constructors are to be used from Java, we probs want to add ` @JvmOverloads` ¯\_(ツ)_/¯ (applies to ones with default values passed in)

**Comment by sollecitom** [comment]
@tlil mind that the API stability check fails because I overloaded 4 functions that happen to be using `net.corda.node.services.config.NodeConfiguration`, which is not public API. We should add those lines to the current API file, because I'm not introducing new types that do not appear in the API, merely overloading. Please confirm? Also @mikehearn FYI.

File: node/src/integration-test/kotlin/net/test/cordapp/v1/FlowCheckpointCordapp.kt (hunk 9 above)
Comment (sollecitom) [reply node/src/integration-test/kotlin/net/test/cordapp/v1/FlowCheckpointCordapp.kt]: Well, no idea, I guess Szymon copied it from somewhere else.

File: node/src/main/kotlin/net/corda/node/cordapp/CordappLoader.kt (hunk 10 above)
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/cordapp/CordappLoader.kt]: By the way, I pulled this interface from the implementation and it's already outlining a few issues (classloader for all cordapps and map flow class -> cordapp, which should be a function of initiating flow name and version header)

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt (hunk 20 above)
Comment (sollecitom) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: No unfortunately, this is PR is already quite big, and while this extra code sucks, we'll have to wait to remove it. It's not easy and quite sensitive, so better do that with a bit more time.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/MutableTestCorDapp.kt (hunk 18 above)
Comment (sollecitom) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/MutableTestCorDapp.kt]: Fixed the second segment to use `File.separator`, but these are not `Path`s, they're `String`s. We do not want complete `Path`s, for they're no good when it comes to pattern matching.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt (hunk 20 above)
Comment (tlil) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: Ok

File: build.gradle [hunk 23]
```diff
@@ -168,7 +168,7 @@ allprojects {
             jvmTarget = "1.8"
             javaParameters = true   // Useful for reflection.
             freeCompilerArgs = ['-Xjvm-default=compatibility']
-            allWarningsAsErrors = true
+            allWarningsAsErrors = project.hasProperty('compilation.allWarningsAsErrors') ? project.property('compilation.allWarningsAsErrors') : false
```
Comment (mikehearn) [inline build.gradle]: This stuff should probably be done in a separate PR, ah well.

**Comment by mikehearn** [comment]
OK to edit the api file.

Ideally we don't edit it outside of releases, but we're being pinned down here by the fact that the API checker is really an ABI checker and we want ABI stability in the core APIs but only need API (source) stability for unit test code.

The checker tool could use some more work to understand more of the stability rules and probably a different way to do the checks (beyond lots of diffs etc).
